src/cpp/src/visual_language/input_embedder.cpp

```diff
@@ -461,73 +461,107 @@ class InputsEmbedderLLaVA : public InputsEmbedder::IInputsEmbedder {
     ) {
         auto text_embeds_shape = text_embeds.get_shape();
-        auto image_embeds_shape = image_embeds.get_shape();
+        size_t text_embeds_seq_length = text_embeds_shape[1];
+        size_t hidden_size = text_embeds_shape[2];
+
+        const int64_t* input_ids_data = input_ids.data<const int64_t>();
+        const float* text_embeds_data = text_embeds.data<const float>();
 
+        size_t num_image_tokens = 0;
+        for (size_t s = 0; s < text_embeds_seq_length; ++s) {
+            if (input_ids_data[s] == image_token_id) {
+                num_image_tokens++;
+            }
+        }
+        auto num_images = image_embeds.size();
         OPENVINO_ASSERT(
-            text_embeds_shape[2] == image_embeds_shape[2],
-            "Incompatible shapes between text_embeds and image_embeds"
+            num_image_tokens == num_images,
+            "Number of image tokens in input_ids different from num_images."
         );
 
-        size_t text_embeds_seq_length = text_embeds_shape[1];
-        size_t hidden_size = text_embeds_shape[2];
-        size_t image_embeds_seq_length = image_embeds_shape[1];
-
-        size_t merged_seq_length = text_embeds_seq_length + (image_embeds_seq_length - 1);
+        size_t total_image_seq_length = 0;
+        for (const auto& single_image_embeds : image_embeds) {
+            OPENVINO_ASSERT(
+                text_embeds_shape[2] == single_image_embeds.get_shape().at(2),
+                "Incompatible shapes between text_embeds and image_embeds"
+            );
+            total_image_seq_length += single_image_embeds.get_shape().at(1);
+        }
+        size_t merged_seq_length = text_embeds_seq_length + total_image_seq_length - num_image_tokens;
 
         ov::Tensor merged_embeds(text_embeds.get_element_type(), {BATCH_SIZE, merged_seq_length, hidden_size});
-
-        const int64_t* input_ids_data = input_ids.data<const int64_t>();
-        const float* text_embeds_data = text_embeds.data<const float>();
-        const float* image_embeds_data = image_embeds.data<const float>();
         float* merged_data = merged_embeds.data<float>();
 
-
         size_t merged_idx = 0;
+        size_t image_idx = 0;
         for (size_t s = 0; s < text_embeds_seq_length; ++s) {
             if (input_ids_data[s] == image_token_id) {
-                for (size_t i = 0; i < image_embeds_seq_length; ++i) {
+                const float* image_embeds_data = image_embeds[image_idx].data<const float>();
+                size_t image_seq_length = image_embeds[image_idx].get_shape()[1];
+
+                for (size_t i = 0; i < image_seq_length; ++i) {
                     std::copy_n(image_embeds_data + i * hidden_size,
-                                hidden_size,
-                                merged_data + merged_idx * hidden_size);
+                            hidden_size,
+                            merged_data + merged_idx * hidden_size);
                     merged_idx++;
```

Comment:
here we have a loop, which copies continuous chunk of memory by `hidden_size` sub-chunks. Can we perform a monolithic copy without a loop? I.e. copy `hidden_size * image_seq_length` elements at once